### PR TITLE
Revise `GamePool` struct and add many capabilities

### PR DESCRIPTION
**Pull request by dmbates**
- Use `PrettyTables.jl` for visualizing a game
- Use `Random.jl` for playing a random game
- Add capability of using entropy to choose the next guess, `guesstype=:entropy`

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,143 +1,431 @@
...
+
+julia> scoreupdate!(primel, [0,1,1,1,2]);
+
+julia> (string(last(primel.guesses)...), last(primel.poolsizes), last(primel.expected), last(primel.entropy))
+("34607", 10, 1.4f0, 2.921928f0)
+
+julia> scoreupdate!(primel, [1,1,1,0,2]);
+
+julia> (string(last(primel.guesses)...), last(primel.poolsizes), last(primel.expected), last(primel.entropy))
+("43867", 2, 1.0f0, 1.0f0)
```
Comment (palday) [inline README.md]: I found this really hard to follow -- could you add some comments? For example, since I know the congruence class trick, I understand why the score update works, but that's not clear from the text so far.

File: README.md [hunk 2]
```diff
@@ -1,143 +1,431 @@
...
-
- Memory estimate: 3.07 KiB, allocs estimate: 50.
- ```
+julia> sum(abs2, wordle.counts) / sum(wordle.counts)  # abs2(x) returns x * x
+61.00086393088553
+```
+
+The next guess is chosen to minimize the expected pool size.
+
+### MAximizing entropy
```
Comment (palday) [inline README.md]:
```suggestion
### Maximizing entropy
```

File: README.md [hunk 3]
```diff
@@ -1,143 +1,431 @@
...
+Measured in bits, the entropy of the `n` probabilities is $-\sum_{i=1}^n p_i\,log_2(p_i)$
+(The negative sign is why the entropy of the 1-choice probability vector `[1.0]` is `-0.0`, as shown above.)
+
+It measures how the probability is dispersed among the possible scores.
+The best case is for each of the `n` possible scores to have probability `1/n` of occurring.In that case, whichever score is returned, there will only be a small number of targets with that score.
+It is not possible to get equal from a starting guess but, sometimes when the target pool is small, a particular guess may be able to split the remaining `k` targets into `k` distinct scores.
+
+In particular, this always occurs when there are only two targets left.
+
+Chosing a guess so as to maximize the entropy provides slightly better performance, on average.
```
Comment (palday) [inline README.md]: This is interesting and it took me a moment to think about why entropy (i.e. expected information) and expected pool size would give different answers. I'm trying to think of something more interesting/expository to say for other readers, but I don't have anything profound other than this is a subtle but cool difference.

File: README.md [hunk 4]
```diff
@@ -1,143 +1,431 @@
...
+│  goner │ 🟩🟩🟩🟩🟩 │        1 │      1.0 │     -0.0 │
+└────────┴────────────┴──────────┴──────────┴──────────┘
+┌────────┬────────────┬──────────┬──────────┬──────────┐
+│  guess │      score │ poolsize │ expected │  entropy │
+│ String │     String │    Int64 │  Float32 │  Float32 │
+├────────┼────────────┼──────────┼──────────┼──────────┤
+│  raise │ 🟫🟩🟫🟫🟫 │     2315 │  61.0009 │  5.87791 │
+│  tangy │ 🟨🟩🟫🟫🟫 │       91 │  7.48352 │  4.03061 │
+│  caput │ 🟨🟩🟫🟫🟨 │       13 │  2.84615 │   2.4997 │
+│  batch │ 🟫🟩🟩🟩🟩 │        5 │      3.4 │ 0.721928 │
```
Comment (palday) [inline README.md]: both of these failures have a weird thing where the entropy drops below 1 bit -- which should be a single guess, before increasing again.

Maybe that's something to note: entropy is dependent on your probability model, which in this case is whether you're looking at the probabilities of score congruence classes or the probabilities of words. Your entropy is on the congruence classes, which means that there might be multiple guesses remaining if they're all in the same class (e.g cover, joker, boxer, foyer are all 02022 here).

I'm not sure my second paragraph makes sense. :thinking:

File: test/runtests.jl [hunk 5]
```diff
@@ -1,6 +1,81 @@
...
+    @test length(first(primel.targetpool)) == 5
+    @test all(reset!(primel).active)
+    @test only(primel.guesses) == ('1', '7', '9', '2', '3')
+    @test only(primel.guessinds) == 826
+    @test only(primel.expected) ≈ 121.54167f0
+    @test only(primel.entropy) ≈ 6.62459f0
+    @test primel.guesstype == :expected
+    @test primel.hardmode
+    playgame!(primel, only(primel.guessinds))   # Got it in one!
+    @test only(primel.scores) == 0xf2
```
Comment (palday) [inline test/runtests.jl:23]: I'm impressed and vaguely terrified that you used hex here :smile:

File: src/play.jl [hunk 6]
```diff
@@ -1,50 +1,132 @@
...
+`S <: Unsigned` is the smallest unsigned integer type that can store values in the
+range `0:(3 ^ N - 1)`, evaluated as `Wordlegames.scoretype(N)`
+
+The fields are:
+
+- `targetpool`: the target pool as a `Vector{NTuple{N,Char}}`
+- `guesspool`: a `Vector{NTuple{N,Char}}` of potential guesses (can be the same as `targets`)
+- `allscores`: a `Matrix{S}` where `allscores[i,j] = score(guesspool[j], targets[i])`
+- `active`: a `BitVector`. The active target pool is `targets[active]`.
+- `counts`: `Vector{Int}` of length `3 ^ N` in which bin counts are accumulated
```
Comment (palday) [inline src/play.jl:20]: why is this `Int` instead of `S`?

File: src/play.jl [hunk 7]
```diff
@@ -1,50 +1,132 @@
...
+- `probs`: `Vector{Float64}` containing `counts ./ sum(counts)`
+- `guesses`: `Vector{NTuple{N,Char}}`
+- `guessinds`: `Vector{Int}` where the indexes into `guesspool` of guesses are stored
```
Comment (palday) [inline src/play.jl:22]: likewise for all the other `Vector{Int}` pieces -- `S` will always be no larger than `Int` and will often be smaller. (If you even need to optimize that much for size.)

File: src/play.jl [hunk 8]
```diff
@@ -1,50 +1,132 @@
...
+struct GamePool{N,S}
+    targetpool::Vector{NTuple{N,Char}}
+    guesspool::Vector{NTuple{N,Char}}
+    allscores::Matrix{S}
+    active::BitVector
     counts::Vector{Int}
-    guess::Ref{Int}
-    expectedpoolsize::Ref{Float64}
-    lastscore::Ref{Int}
+    probs::Vector{Float32}
```
Comment (palday) [inline src/play.jl]: This is Float64 in the docstring.

I would probably use Float64 for all of these things -- at most you're looking at 3 * 32 = 96 bytes of waste and Float32 will often not be more efficient on 64 bit architectures (due to alignment and register design). If we were going for GPUs, then that would be a different story. :smile:

File: src/play.jl [hunk 9]
```diff
@@ -54,57 +136,139 @@ function bincounts!(
...
+`probs` is assumed to be a discrete probability distribution.  That is the conditions
+`all(0 .≤ probs .≤ 1) && sum(probs) ≈ 1` are assumed (but not checked).
 """
-function reset!(gp::GamePool)
-    fill!(gp.active, true)
-    gp.guess[] = gp.firstguess
-    gp.expectedpoolsize[] = gp.firstexpectedpoolsize
-    return gp
+function entropybase2(probs::AbstractVector{<:AbstractFloat})
+    return -sum(x -> iszero(x) ? zero(x) : x * log2(x), probs)
```
Comment (palday) [inline src/play.jl]: Using StatsFuns.jl, you could write this as
```suggestion
    return -sum(x -> xlogx(x) / logtwo, probs)
```
though I don't know if the extra division will be faster or slower than the branch, esp since there's probably a branch in `xlogx`

File: test/runtests.jl (hunk 5 above)
Comment (dmbates) [reply test/runtests.jl:23]: The default printing of a `GamePool` struct in the REPL shows the scores as the `S` type and I just copied the hex representation. :shrug:

File: src/play.jl (hunk 6 above)
Comment (dmbates) [reply src/play.jl:20]: `S` is the smallest `Unsigned` type that can represent a score.  The maximum value in `counts` is the length of the target pool, which can be bigger than the maximum score.

Comment (palday) [reply src/play.jl:20]: Ah, yeah makes sense :smile:

File: src/play.jl (hunk 8 above)
Comment (dmbates) [reply src/play.jl]: You're right.  I think I may delete the `probs` field and make it a property which can be calculated on the fly, using your next suggestion to write `entropybase2` as a function of `counts`
```jl
function entropybase2(counts::AbstractVector(<:Real))
    countsum = sum(counts)
    return sum(x -> xlogx(x / countsum) / logtwo, counts)
end
```

File: src/play.jl (hunk 9 above)
Comment (dmbates) [reply src/play.jl]: There is a branch in `xlogx` but that's not important.  Your code is cleaner.

**Comment by dmbates** [comment]
Thanks for the comments @palday .  I have a couple of questions.

- Should the default `guesstype` be `:entropy`?  It seems to be more efficient so why not make it the default?
- How would I go about making the `guesstype` a parameter of the `GamePool` struct instead of a field?  I want to use dispatch on `updateguess!` methods instead of a branch in the function.

**Comment by palday** [comment]
> Thanks for the comments @palday . I have a couple of questions.
> 
>     * Should the default `guesstype` be `:entropy`?  It seems to be more efficient so why not make it the default?

Seems reasonable.

>     * How would I go about making the `guesstype` a parameter of the `GamePool` struct instead of a field?  I want to use dispatch on `updateguess!` methods instead of a branch in the function.

Define new singleton types and dispatch on them (and this also gives you error handling):

```julia
abstract type GuessType end

struct MinimizeEntropy <: GuessType end

struct MinimizePoolSize  <: GuessType end
```

The compiler is now smart enough to optimize all that quite well (via inlining and constant propagation), and this replaces the old `Val` types in lot of contexts -- the LinearAlgebra stdlib is being rewritten to take advantage of this for e.g. Cholesky and QR factorizations.
